engine/pkg/election/elector.go

```diff
@@ -247,31 +220,55 @@ func (e *Elector) release(ctx context.Context, removeSelf bool) error {
 	ctx, cancel := context.WithTimeout(ctx, defaultReleaseTimeout)
 	defer cancel()
 
-	s := e.config.Storage
+	return e.updateRecord(ctx, func(record *Record) error {
+		if record.LeaderID == e.config.ID {
+			record.LeaderID = ""
+		}
+		if removeSelf {
+			for i, m := range record.Members {
+				if m.ID == e.config.ID {
+					record.Members = append(record.Members[:i], record.Members[i+1:]...)
+					break
+				}
+			}
+		}
+		return nil
+	})
+}
 
-	record, err := s.Get(ctx)
-	if err != nil {
-		return errors.Trace(err)
+func (e *Elector) updateRecord(ctx context.Context, f func(*Record) error) error {
+	// Divide 2 is for more retries.
+	backoffBaseDelayInMs := int64(e.config.RenewInterval/time.Millisecond) / 2
+	// Make sure the retry delay is less than the deadline, otherwise the retry has no chance to execute.
+	backoffMaxDelayInMs := int64(e.config.RenewDeadline/time.Millisecond) / 2
+	if deadline, ok := ctx.Deadline(); ok {
+		backoffMaxDelayInMs = int64(deadline.Sub(time.Now())/time.Millisecond) / 2
 	}
-	e.setObservedRecord(record)
 
-	if record.LeaderID == e.config.ID {
-		record.LeaderID = ""
-	}
-	if removeSelf {
-		for i, m := range record.Members {
-			if m.ID == e.config.ID {
-				record.Members = append(record.Members[:i], record.Members[i+1:]...)
-				break
-			}
+	return retry.Do(ctx, func() error {
+		s := e.config.Storage
+		record, err := s.Get(ctx)
+		if err != nil {
+			return errors.Trace(err)
 		}
-	}
+		e.setObservedRecord(record)
 
-	if err := s.Update(ctx, record); err != nil {
-		return errors.Trace(err)
-	}
-	e.setObservedRecord(record)
-	return nil
+		if err := f(record); err != nil {
+			return errors.Trace(err)
+		}
+
+		if err := s.Update(ctx, record); err != nil {
+			return errors.Trace(err)
+		}
+		e.setObservedRecord(record)
+		return nil
+	}, retry.WithBackoffBaseDelay(backoffBaseDelayInMs),
+		retry.WithBackoffMaxDelay(backoffMaxDelayInMs),
+		retry.WithIsRetryableErr(func(err error) bool {
+			log.Debug("failed to update record", zap.Error(err))
```

Comment:
There is a info log before, `log.Info("renew lease failed, retry after random delay",`
Is it better to use Info level here

Reply:
Addressed. Info for conflict and Warn for other errors.